### ISSUE TITLE
[bug] Transformer Engine fused cross-entropy zeroes MTP acceptance metrics

### ISSUE THREAD

**Issue by milank94** [NVIDIA-NeMo/Megatron-Bridge#5947]
### Problem

`mtp_{k}_acceptance_rate` is reported as exactly zero for every MTP depth and training step when MTP training is combined with:

```bash
cross_entropy_loss_fusion: true
cross_entropy_fusion_impl: "te"
```

MTP losses otherwise train normally, making this a silent metrics failure rather than an obvious training failure.

Initial analysis suggests an ordering interaction between `process_mtp_loss` and Transformer Engine’s fused cross-entropy path, which may modify the tensors needed to calculate the acceptance metrics.

### Minimal repro

1. Start from nvcr.io/nvidia/nemo:26.08.00.
2. Launch an MTP-enabled SFT run through megatron.bridge.training.finetune.
3. Use the following relevant configuration:

```shell
mtp_num_layers: 5
mtp_use_repeated_layer: true
mtp_loss_scaling_factor: 0.2
calculate_per_token_loss: true
cross_entropy_loss_fusion: true
cross_entropy_fusion_impl: "te"
```

4. Observe that all mtp_{k}_acceptance_rate metrics remain at zero while the MTP losses continue to train normally.

Reproduced with Qwen/Qwen3.5-0.8B using the qwen3_vl_step forward step.

### Expected behavior

MTP acceptance metrics should be calculated and reported correctly when Transformer Engine fused cross-entropy is enabled. The fused loss path should not overwrite or modify values required by `process_mtp_loss` before metric calculation.

### Affected area

area:training

### Regression?

Not sure

### Environment

- Container: `nvcr.io/nvidia/nemo:26.08.00`
- Megatron-Bridge: `v0.6.0`
- Megatron-Core: `0.19.0`
- Transformer Engine: `2.16.0+4220403e`
- GPU setup: 1 node, 8x H200
- Model: Qwen/Qwen3.5-0.8B dense (~151K vocabulary)
- Workload: SFT
- Sequence length: 131,072
- Micro batch size: 1
- Global batch size: 16
- Sequence parallelism: enabled
- Precision: BF16 with tensorwise FP8
- Parallelism: TP=2, CP=2, PP=1, DP=2

### Logs

No traceback is produced. Training and MTP losses proceed normally, but all mtp_{k}_acceptance_rate metrics remain exactly zero. Raw metric output can be added if needed.

**Comment by yaoyu-33** [comment]
We isolated this with a targeted one-GPU runtime reproducer that calls the real MCore `process_mtp_loss` path. This is an MCore call-ordering bug exposed by the destructive Transformer Engine fused-CE implementation, rather than a model-, sequence-length-, or parallelism-specific failure.

### Root cause

`process_mtp_loss` currently does the following in order:

1. `compute_language_model_loss(mtp_labels, mtp_logits)`
2. `_compute_mtp_acceptance_counts(mtp_logits, ...)`

This ordering was introduced with the acceptance metrics in [NVIDIA/Megatron-LM#3458](https://github.com/NVIDIA/Megatron-LM/pull/3458). It is present in the MCore revision pinned by Bridge v0.6.0 and remains present in the MCore revision currently pinned on Bridge `main`.

Transformer Engine fused CE through 2.17.1 reuses the input-logits storage for the precomputed gradient, effectively replacing each row with `softmax(logits) - one_hot(label)` during forward. The target entry becomes negative while non-target entries are nonnegative, so the later acceptance `argmax` cannot select the target. The loss remains correct because it is computed before the input buffer is overwritten. The underlying destructive-input behavior is also tracked by [NVIDIA/TransformerEngine#2899](https://github.com/NVIDIA/TransformerEngine/issues/2899).

### Targeted A/B result

The synthetic input was constructed with 100% pre-loss acceptance. Two MTP depths were evaluated with BF16 logits:

| CE path | Logit elements changed after CE | Acceptance after CE | MTP losses |
| --- | ---: | ---: | ---: |
| TE fused | 192/192 per depth | 0/10 and 0/8 | ~0.00528 and ~0.00527 |
| Native fused | 0/192 | 10/10 and 8/8 | equivalent |
| Unfused | 0/192 | 10/10 and 8/8 | equivalent |

For the TE path, every inspected target value changed from `8.0` to `-0.00390625`; labels were unchanged. The recorded event order was CE return followed by acceptance calculation. This reproduces the silent failure without the original model or 131K sequence length. Runtime validation covered TE 2.17.0 and 2.17.1; source comparison confirms the same destructive path in the reported TE 2.16 revision.

### Workarounds

For BF16 training on the 26.08 stack, use native fused CE:

```yaml
cross_entropy_loss_fusion: true
cross_entropy_fusion_impl: native
```

Alternatively, disable CE fusion:

```yaml
cross_entropy_loss_fusion: false
```

MCore already rejects TE fused CE from its CLI validation due to known stability issues ([NVIDIA/Megatron-LM#5115](https://github.com/NVIDIA/Megatron-LM/pull/5115)); programmatic configs only warn after [NVIDIA/Megatron-LM#5162](https://github.com/NVIDIA/Megatron-LM/pull/5162), which is why Bridge callers can still enter this combination.

[NVIDIA/TransformerEngine#3193](https://github.com/NVIDIA/TransformerEngine/pull/3193), released in TE 2.18, stops reusing the logits buffer and therefore avoids this specific symptom. However, the robust MCore fix is to compute acceptance before invoking CE. That avoids a long-sequence logits clone and does not rely on any CE backend preserving its input; native/unfused paths can also reuse storage when logits are already FP32.

**Comment by huvunvidia** [comment]
Hi @milank94 ,
Just check-in to see if @yaoyu-33 's suggestion fixes your problem.
If so or you want to leave this issue for later, we can close it temporarily and reopen it later if needed.

**Comment by linmuchuiyang** [comment]
Additional data point from H100 (nemo:26.08, Megatron-Bridge v0.6.0, Megatron-LM dev `43e45e13`, TE 2.17.1): reproduced with DeepSeek-V4-Pro and DeepSeek-V4-Flash SFT, `cross_entropy_fusion_impl=te`. `mtp_1_acceptance_rate` is 0.0 at every step in both runs, while the MTP top-1 accuracy recomputed from a **copy** of `mtp_logits` taken before `compute_language_model_loss` is 0.48 (same as the LM head's accuracy against t+1). This matches @yaoyu-33's analysis above (TE fused CE overwrites the logits buffer in place, acceptance is computed afterwards on the same tensor). Computing `_compute_mtp_acceptance_counts` before the loss, or on a clone when the TE path is active, fixes the metric in our probe. Happy to open the Megatron-LM PR if nobody has it in flight.

**Comment by dundysm** [comment]
i will open a megatron-lm pr that computes mtp acceptance before the te fused-ce path can overwrite logits in place (acceptance ordering fix in `process_mtp_loss`).

**Comment by dundysm** [comment]
opened megatron-lm pr for the acceptance ordering fix: https://github.com/NVIDIA/Megatron-LM/pull/7623
